lib/elixir/test/elixir/kernel/expansion_test.exs

```diff
@@ -2602,7 +2624,7 @@ defmodule Kernel.ExpansionTest do
         expand(code)
       end
 
-      message = ~r"size in bitstring expects an integer or a variable as argument, got: foo()"
+      message = ~r"cannot find or invoke local foo/0 inside guard"
```

Comment:
This error message can be confusing now, since it says guard but it isn't really a guard? However, if we are in a guard, we likely shouldn't try to invoke a function. So can you please do these following additional changes? Change this:

https://github.com/elixir-lang/elixir/blob/d2f899b04d03d2aa62196cec927545478fddd716/lib/elixir/src/elixir.hrl#L1

to this:

    -define(key(M, K), map_get(K, M)).

And then change this:

https://github.com/elixir-lang/elixir/blob/d2f899b04d03d2aa62196cec927545478fddd716/lib/elixir/src/elixir_expand.erl#L387

To this:

                warn when ?key(E, context) == nil ->

Then the error message should just say undefined variable. :)

Reply:
@josevalim I tried these modifications but there was no change in the error message for this specific test case :(

Also, it failed in another test when the `prematch` is `warn` and the context is `guard`, so the match fails. We would need a default case statement, or maybe matching `warn` without conditionals in a separate clause, but I don't quite follow the intention behind your changes so I don't know (just a few days messing with the codebase, bear with me 😄). 

```
  1) test locals in guards (Kernel.ExpansionTest)
     lib/elixir/test/elixir/kernel/expansion_test.exs:418
     ** (MatchError) no match of right hand side value: true
     code: assert clean_meta(expand(code), [:import, :context]) == expanded_code
     stacktrace:
       ...
       lib/elixir/test/elixir/kernel/expansion_test.exs:2827: Kernel.ExpansionTest.expand/1
       lib/elixir/test/elixir/kernel/expansion_test.exs:421: (test)
```

But again, this change is not modifying the error message that says `cannot find or invoke local foo/0 inside guard`. It keeps returning that.

Moreover, there are other cases where the error message refers to clauses, it seems hard to avoid without further thinking/work. See another example:

```ex
iex(2)> <<ff::size(foo.())>> = <<1,2,3>>
** (CompileError) iex:2: invalid expression in guard, anonymous call is not allowed in guards. To learn more about guards, visit: https://hexdocs.pm/elixir/patterns-and-guards.html
```

I agree with you that referring to `guards` in error messages is not natural in the context of the size of bitstrings, so we probably need to keep thinking about how to deal with it.
The only idea I have in mind is to add another possible value for the `context` (the thing that is `nil`, `match` or `guard`), so we can differentiate this special case when forming errors in `exilir_expand.erl`. It's kind of polluting the existing code with an additional case that behaves almost exactly like the `guard` case, but maybe it's our only choice? Sharing to know your opinion. If it's feasible I can spend some time trying it.

Reply:
Good points. I think in this case then the best option is to add a field here:

https://github.com/elixir-lang/elixir/blob/main/lib/elixir/src/elixir.hrl#L8

It can be a boolean for now: `bitsize=true|false`. And then add a `elixir_utils:context()` that returns the actual context? Unfortunately we cannot change `context = guard` without breaking existing code.

Reply:
I like the idea.. will try it 👍